blssig/src/verifier/mod.rs

```diff
@@ -0,0 +1,169 @@
+// Copyright 2019-2024 ChainSafe Systems
+// SPDX-License-Identifier: Apache-2.0, MIT
+
+use bls_signatures::{PublicKey, Serialize, Signature, verify_messages};
+use filecoin_f3_gpbft::PubKey;
+use filecoin_f3_gpbft::api::Verifier;
+use hashlink::LruCache;
+use parking_lot::RwLock;
+use thiserror::Error;
+
+use crate::bdn::BDNAggregation;
+
+#[cfg(test)]
+mod tests;
+
+#[derive(Error, Debug, PartialEq)]
+pub enum BLSError {
+    #[error("empty public keys provided")]
+    EmptyPublicKeys,
+    #[error("invalid public key length: expected {BLS_PUBLIC_KEY_LENGTH} bytes, got {0}")]
+    InvalidPublicKeyLength(usize),
+    #[error("failed to deserialize public key")]
+    PublicKeyDeserialization,
+    #[error("invalid signature length: expected {BLS_SIGNATURE_LENGTH} bytes, got {0}")]
+    InvalidSignatureLength(usize),
+    #[error("failed to deserialize signature")]
+    SignatureDeserialization,
+    #[error("BLS signature verification failed")]
+    SignatureVerificationFailed,
+    #[error("mismatched number of public keys and signatures: {pub_keys} != {sigs}")]
+    LengthMismatch { pub_keys: usize, sigs: usize },
+}
+
+/// BLS signature verifier using BDN aggregation scheme
+///
+/// This verifier implements the same scheme used by `go-f3/blssig`, with:
+/// - BLS12_381 curve
+/// - G1 for public keys, G2 for signatures  
+/// - BDN aggregation for rogue-key attack prevention
+/// - Point cache for performance optimization
+pub struct BLSVerifier {
+    /// Cache for deserialized public key points to avoid expensive repeated operations
+    point_cache: RwLock<LruCache<Vec<u8>, PublicKey>>,
```

Comment:
I'd change the signatures of the `Verifier` trait methods to use  strongly-typed public key and signature to avoid conversions and such LRU perf optimization

Reply:
I followed the existing `trait Verifier`, which follows `type Verifier interface` from `go-f3`, who's using untyped params for all functions.

However, I agree that the `bdn` module can/should receive the typed data, so I changed that here: https://github.com/ChainSafe/rust-f3/pull/27/commits/3608b91212c4f5e42a7b6b430f06d5548cf1337e

I'm not convinced though that `Verifier` should do the same, since it will merely push the validity checks, the deserialization and the caching to a higher layer, given that the data source is serialized.